Add note about ghcr.io for HOMEBREW_DOCKER_REGISTRY_TOKEN

azatoth opened the pull request:
It's not clear that when using HOMEBREW_DOCKER_REGISTRY_TOKEN directly against ghcr.io that you should use your PAT base64-encoded as the bearer token.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -165,8 +165,20 @@ module EnvConfig
-        description: "Use this bearer token for authenticating with a Docker registry proxying GitHub Packages. " \
-                     "Preferred over `HOMEBREW_DOCKER_REGISTRY_BASIC_AUTH_TOKEN`.",
+        description:  <<~EOF,
+          Use this bearer token for authenticating with a Docker registry proxying GitHub Packages.
+          Preferred over `HOMEBREW_DOCKER_REGISTRY_BASIC_AUTH_TOKEN`.
+
+              *Note:* when authenticating against ghcr.io, \
+              the `HOMEBREW_DOCKER_REGISTRY_TOKEN` must be a base64 \
+              encoded GitHub Personal Access Token (PAT).
```
Not convinced this description makes that much sense after this? If you're "authenticating against ghcr.io" is this "a Docker registry proxying GitHub Packages"? It doesn't sound like it. Maybe the whole description needs a bigger overhaul?

azatoth replied:
Yea, this is related to the underlying issue I had, trying to access a private tap hosted on a private GitHub repo. And while optimally, there should be a way to be able to assign different tokens for different taps, at the moment this environment variable is the only way to address brew to authenticate again ghcr.io.

MikeMcQuaid replied:
I wonder if this should default to `HOMEBREW_GITHUB_API_TOKEN` and/or read from the repository URL like https://github.com/Homebrew/brew/pull/16649. Avoiding the user from having to manually run `base64` seems ideal.

azatoth replied:
> should default to `HOMEBREW_GITHUB_API_TOKEN`

While it's highly possibly it's the same token, conceptually it's wrong as the docker registry doesn't have anything to do with the API and vice versa. Wouldn't defaulting to `HOMEBREW_GITHUB_PACKAGES_TOKEN` be more logical (which in itself then might need to be promoted out from development towards more general use)?

> and/or read from the repository URL

I assume it makes sense to allow it to read from the URL, even though having the token in the URL isn't something I personally would ever do.

MikeMcQuaid replied:
> Wouldn't defaulting to `HOMEBREW_GITHUB_PACKAGES_TOKEN` be more logical (which in itself then might need to be promoted out from development towards more general use)?

Yup, this would make sense!

MikeMcQuaid replied:
@azatoth Sorry for the confusion here, for more concrete ask here either:
- default to `HOMEBREW_GITHUB_API_TOKEN` and do the Base64 internally inside of Homebrew
or:
- update the description better here as `ghcr.io` is not "a Docker registry proxying GitHub Packages" but instead GitHub Packages itself.

azatoth replied:
> * default to `HOMEBREW_GITHUB_API_TOKEN`

Don't you mean `HOMEBREW_GITHUB_PACKAGES_TOKEN`?

MikeMcQuaid replied:
No, that's another option we could use but if `HOMEBREW_GITHUB_API_TOKEN` is setup correctly we should be able to Base64 it and use it accordingly.

azatoth replied:
> No, that's another option we could use but if `HOMEBREW_GITHUB_API_TOKEN` is setup correctly we should be able to Base64 it and use it accordingly.

I think I'm a bit confused, I thought the API token was for accessing the API and PACKAGES token to access packages in the event you would want to have different tokes with limited permissions in each

MikeMcQuaid replied:
`HOMEBREW_GITHUB_PACKAGES_TOKEN` is currently only used for uploading packages.

azatoth replied:
> * default to `HOMEBREW_GITHUB_API_TOKEN` and do the Base64 internally inside of Homebrew

I decided to try this route;
I did rebase against master, but I don't know what your policy for PRs actually is in this regard.

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -165,8 +165,20 @@ module EnvConfig
       HOMEBREW_DOCKER_REGISTRY_TOKEN:            {
-        description: "Use this bearer token for authenticating with a Docker registry proxying GitHub Packages. " \
-                     "Preferred over `HOMEBREW_DOCKER_REGISTRY_BASIC_AUTH_TOKEN`.",
+        description:  <<~EOF,
```
```suggestion
        description: <<~EOS,
```

azatoth replied:
fixed in c2c0c24546

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
````diff
@@ -165,8 +165,20 @@ module EnvConfig
+          ```bash
+          export HOMEBREW_DOCKER_REGISTRY_TOKEN=$(base64<<<$(git config github.token))
+          ```
+        EOF
````
```suggestion
        EOS
```

azatoth replied:
fixed in c2c0c24546

MikeMcQuaid reviewed (commented):
Thanks @azatoth! Definitely think this needs changing, just let's riff off exactly what to.

Bo98 commented:
(Prior discussion for context: https://github.com/orgs/Homebrew/discussions/5171)

MikeMcQuaid reviewed (commented):
Would like to see comments addressed or further discussion, thanks!

azatoth commented:
> Would like to see comments addressed or further discussion, thanks!

I might have misunderstood you, but I thought I had addressed your comments, and I don't know what more I can say to further the issue.

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -239,12 +242,20 @@ module EnvConfig
-                     "<https://docs.github.com/en/rest/overview/rate-limits-for-the-rest-api>" \
-                     "\n\n    *Note:* Homebrew doesn't require permissions for any of the scopes, but some " \
-                     "developer commands may require additional permissions.",
+        description: <<~EOS,
```
Please put this all one one line too.

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -239,12 +242,20 @@ module EnvConfig
+              *Note:* If set, this token will also be used to authenticating against `ghcr.io` unless
+                      `HOMEBREW_DOCKER_REGISTRY_TOKEN` or `HOMEBREW_DOCKER_REGISTRY_BASIC_AUTH_TOKEN` has been set.
+                      In this case, the token will require the `packages` permission to be set.
+                      `HOMEBREW_GITHUB_PACKAGES_TOKEN` is not used for this permission, but only for uploading packages.
```
```suggestion
```

Remove this and move to the `HOMEBREW_GITHUB_PACKAGES_TOKEN` description (if there is one).

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -239,12 +242,20 @@ module EnvConfig
+              *Note:* Homebrew doesn't require permissions for any of the scopes,
+                      but some developer commands may require additional permissions.
+
+              *Note:* If set, this token will also be used to authenticating against `ghcr.io` unless
```
Love this, great work!

MikeMcQuaid reviewed (commented):
Great work @azatoth, almost done here!

EricFromCanada commented on Library/Homebrew/env_config.rb:
```diff
@@ -239,12 +242,20 @@ module EnvConfig
+          Use this personal access token for the GitHub API, for features such as `brew search`.
+          You can create one at <https://github.com/settings/tokens>.
+          If set, GitHub will allow you a greater number of API requests.
+          For more information, see: "<https://docs.github.com/en/rest/overview/rate-limits-for-the-rest-api>"
```
```suggestion
          For more information, see: <https://docs.github.com/en/rest/overview/rate-limits-for-the-rest-api>
```
